### PR TITLE
Compiles and run on Scala 2.12, latest version of Ensime (2.0.0-M4). Added settings to change server log level from VSCode

### PR DESCRIPTION
**Pull request by fede0664**
- Scala 2.12: The challenge was upgrading scala-json-rpc to version 2.0.0, required a few changes on Commands.scala and JsonRpcUtils.scala
- Ensime (2.0.0-M4): implemented changes suggested at #50. The ActorInitializationException seems to be gone, but Ensime still has some irrecoverable fails, for example when getting the definition of a Manifest. #37 still not fixed.
- Settings change: Added a flag in VSCode settings that is sent to Coursier to change the default logging level for org.github.dragos.vscode on loogback.groovy. The goal of this is to limit the verbosity of the log on production, log files grow quickly to a several GBs in DEBUG level.

### CODE REVIEW

**Comment by dragos** [comment]
I just noticed that Document Symbols stopped working. I see these errors in the log, perhaps related to the upgrade to scala-json-rpc 2.0?

```
[2017-09-11 14:34:55,435] DEBUG org.github.dragos.vscode.EnsimeLanguageServer - got back: List(StructureViewMember(class,Hello,OffsetSourcePosition(RawFile(/Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala),8),List(StructureViewMember(def,foo,OffsetSourcePosition(RawFile(/Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala),22),List()), StructureViewMember(type,X,OffsetSourcePosition(RawFile(/Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala),135),List()))))
[2017-09-11 14:34:55,435] DEBUG org.github.dragos.vscode.EnsimeLanguageServer - got back: List(StructureViewMember(class,Hello,OffsetSourcePosition(RawFile(/Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala),8),List(StructureViewMember(def,foo,OffsetSourcePosition(RawFile(/Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala),22),List()), StructureViewMember(type,X,OffsetSourcePosition(RawFile(/Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala),135),List()))))
[2017-09-11 14:34:55,437] INFO langserver.core.Connection - Result:DocumentSymbolResult(List(SymbolInformation(Hello,5,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(2,6),Position(2,11))),None), SymbolInformation(foo,6,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(3,6),Position(3,9))),Some(Hello)), SymbolInformation(X,11,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(11,7),Position(11,8))),Some(Hello))))
[2017-09-11 14:34:55,437] INFO langserver.core.Connection - Result:DocumentSymbolResult(List(SymbolInformation(Hello,5,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(2,6),Position(2,11))),None), SymbolInformation(foo,6,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(3,6),Position(3,9))),Some(Hello)), SymbolInformation(X,11,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(11,7),Position(11,8))),Some(Hello))))
[2017-09-11 14:34:55,441] ERROR langserver.core.Connection - ResultResponse.write:DocumentSymbolResult(List(SymbolInformation(Hello,5,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(2,6),Position(2,11))),None), SymbolInformation(foo,6,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(3,6),Position(3,9))),Some(Hello)), SymbolInformation(X,11,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(11,7),Position(11,8))),Some(Hello))))
[2017-09-11 14:34:55,441] ERROR langserver.core.Connection - ResultResponse.write:DocumentSymbolResult(List(SymbolInformation(Hello,5,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(2,6),Position(2,11))),None), SymbolInformation(foo,6,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(3,6),Position(3,9))),Some(Hello)), SymbolInformation(X,11,Location(file:///Users/dragos/workspace/oss/test-code/src/main/scala/Hello.scala,Range(Position(11,7),Position(11,8))),Some(Hello))))
```

File: languageserver/src/main/scala/langserver/utils/JsonRpcUtils.scala [hunk 1]
```diff
@@ -34,8 +34,8 @@ object JsonRpcUtils {
    *         )
    * }}}
    */
-   def valueFormat[A, B: Format](apply: B => A)(unapply: A => B): Format[A] = new Format[A] {
+   def valueFormat[A, B: Format](apply: B => A)(unapply: A => B): OFormat[A] = new OFormat[A] {
     override def reads(json: JsValue) = Reads.of[B].reads(json).map(apply(_))
-    override def writes(o: A) = Writes.of[B].writes(unapply(o))
+    override def writes(o: A) = Writes.of[B].writes(unapply(o)).as[JsObject]
```
Comment (dragos) [inline languageserver/src/main/scala/langserver/utils/JsonRpcUtils.scala]: This seems to be the cause of the exception I reported earlier. In the case of `DocumentSymbolResult` the underlying type is a sequence, not a Json object.

Comment (fede0664) [reply languageserver/src/main/scala/langserver/utils/JsonRpcUtils.scala]: If that's the case we'll need a similar workaround like the one I implemented for the Locations for "textDocument/definition". It seems that the latest play-json-rcp Message.MessageFormats only accepts JsObject, can't deal with JsValue. I'll take a look later today.
Can you add a test case to include this case?

Comment (fede0664) [reply languageserver/src/main/scala/langserver/utils/JsonRpcUtils.scala]: I pushed the changes to fix the issue with "textDocument/definition". I had to use the same workaround used for the Seq[Location]. Changed JsonRpcUtils.valueFormat to work only with JsObject.

**Comment by dragos** [comment]
Fantastic, thanks a lot! Merging this.

**Comment by jvican** [comment]
Yay!! You cannot imagine the atrocities I've done to my 2.12 build to support the vscode integration for scalac-performance here: https://github.com/scalacenter/scalac-profiling.

**Comment by dhpiggott** [comment]
Hey. scala-json-rpc author here. I see now from looking over this (found it via https://twitter.com/jaguarul/status/907514621728260096) that this was a regression in scala-json-rpc 2.0.0.

I've just fixed it via https://github.com/dhpcs/scala-json-rpc/pull/48 and released the fixed version as 2.0.1, so it should be possible now to update dragos-vscode-scala and remove these workarounds.

**Comment by dragos** [comment]
Thank you @dhpiggott! Open-source at its finest!

**Pull request merged**